tests/server/api/api_test.go

```diff
@@ -264,6 +264,126 @@ func (s *testMiddlewareSuite) TestAuditMiddleware(c *C) {
 	c.Assert(failpoint.Disable("github.com/tikv/pd/server/api/addAuditMiddleware"), IsNil)
 }
 
+func (s *testMiddlewareSuite) TestRateLimitMiddleware(c *C) {
+	c.Assert(failpoint.Enable("github.com/tikv/pd/server/api/addRateLimitMiddleware", "return(true)"), IsNil)
+	leader := s.cluster.GetServer(s.cluster.GetLeader())
+	req, _ := http.NewRequest("POST", leader.GetAddr()+"/pd/api/v1/admin/ratelimit-middleware?enable=true", nil)
+	resp, err := dialClient.Do(req)
+	c.Assert(err, IsNil)
+	resp.Body.Close()
+	c.Assert(leader.GetServer().IsRateLimitMiddlewareEnabled(), Equals, true)
+
+	req, _ = http.NewRequest("GET", leader.GetAddr()+"/pd/api/v1/fail/", nil)
+	resp, err = dialClient.Do(req)
+	c.Assert(err, IsNil)
+	_, err = io.ReadAll(resp.Body)
+	resp.Body.Close()
+	c.Assert(err, IsNil)
+	c.Assert(resp.StatusCode, Equals, http.StatusOK)
+	c.Assert(resp.Header.Get("rate-limit"), Equals, "rate-limit")
+
+	req, _ = http.NewRequest("POST", leader.GetAddr()+"/pd/api/v1/admin/ratelimit-middleware?enable=false", nil)
+	resp, err = dialClient.Do(req)
+	c.Assert(err, IsNil)
+	resp.Body.Close()
+	c.Assert(leader.GetServer().IsRateLimitMiddlewareEnabled(), Equals, false)
+
+	req, _ = http.NewRequest("GET", leader.GetAddr()+"/pd/api/v1/fail/", nil)
+	resp, err = dialClient.Do(req)
+	c.Assert(err, IsNil)
+	_, err = io.ReadAll(resp.Body)
+	resp.Body.Close()
+	c.Assert(err, IsNil)
+	c.Assert(resp.StatusCode, Equals, http.StatusOK)
+	c.Assert(resp.Header.Get("rate-limit"), Equals, "")
+	c.Assert(failpoint.Disable("github.com/tikv/pd/server/api/addRateLimitMiddleware"), IsNil)
+}
+
+func (s *testMiddlewareSuite) TestRateLimiterDontAllowRequset(c *C) {
+	c.Assert(failpoint.Enable("github.com/tikv/pd/pkg/ratelimiter/dontAllowRatelimiter", "return(true)"), IsNil)
+	leader := s.cluster.GetServer(s.cluster.GetLeader())
+	req, _ := http.NewRequest("POST", leader.GetAddr()+"/pd/api/v1/admin/ratelimit-middleware?enable=true", nil)
+	resp, err := dialClient.Do(req)
+	c.Assert(err, IsNil)
+	resp.Body.Close()
+	c.Assert(leader.GetServer().IsRateLimitMiddlewareEnabled(), Equals, true)
+
+	req, _ = http.NewRequest("GET", leader.GetAddr()+"/pd/api/v1/fail/", nil)
+	resp, err = dialClient.Do(req)
+	c.Assert(err, IsNil)
+	_, err = io.ReadAll(resp.Body)
+	resp.Body.Close()
+	c.Assert(err, IsNil)
+	c.Assert(resp.StatusCode, Equals, http.StatusTooManyRequests)
```

Comment:
I think you can check if the rate limit can actually work instead of using the failpoint.

Reply:
That's right.  I will change it after merging https://github.com/tikv/pd/pull/4501